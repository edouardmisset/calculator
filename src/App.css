:root {
  font-family: Inter, system-ui, Avenir, Helvetica, Arial, sans-serif;
  line-height: 1.5;
  font-weight: 400;

  color-scheme: light dark;
  color: rgba(255, 255, 255, 0.87);
  background-color: #242424;

  --main-color: hsl(26, 95%, 48%);
  --green: hsl(74, 38%, 62%);
  --dark-green: hsl(106, 75%, 55%);
  --white: hsl(0, 0%, 95%);
  --grey: hsl(0, 0%, 41%);
  --dark-grey: hsl(0, 0%, 18%);
  --black: hsl(0, 0%, 8%);
  --orange: hsl(16, 100%, 63%);
  --red: rgb(185, 49, 15);
  --yellow: rgb(255, 237, 72);
  --blue: hsl(214, 87%, 59%);
  --border-radius: 1rem;

  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: 100%;
}

html,
body,
h1,
h2 {
  margin: 0;
  padding: 0;
}

body {
  display: flex;
  justify-content: center;
  align-items: center;
  background: linear-gradient(to bottom right, var(--blue), var(--dark-green));
  min-width: 500px;
  height: 100dvh;
  min-height: 100dvh;
}

#root {
  max-width: 1280px;
  margin: 0 auto;
  text-align: center;
  height: 100%;
}

h1 {
  font-size: 5em;
  line-height: 1;
}

header {
  font-family: monospace;
  font-weight: 900;

  margin-block-start: 1em;
  margin-block-end: 2em;

  color: var(--orange); /* Fallback solid color */
  -webkit-text-fill-color: transparent; /* Webkit Disable Fallback */

  /* Text color gradient properties */
  background: -webkit-linear-gradient(var(--red), var(--yellow));
  -webkit-background-clip: text;
  -webkit-text-fill-color: transparent;
  background-clip: text;
}

.calculator {
  margin: auto;
  display: grid;
  grid-template-areas:
    'screen screen screen screen'
    'screen screen screen screen'
    'clear inverse divide multiply'
    'seven eight nine add'
    'four five six subtract'
    'one two three equal'
    'zero zero dot equal';
  max-width: 500px;
  border-radius: var(--border-radius);
}

.screen {
  border: none;
  border-start-start-radius: var(--border-radius);
  border-start-end-radius: var(--border-radius);
  grid-area: screen;
  text-align: right;
  color: var(--black);
  font-size: 2rem;
  padding: 2rem;
  min-height: 4rem;
  font-family: monospace;
}

.equal {
  grid-area: equal;
  border-end-end-radius: var(--border-radius);
}

.zero {
  grid-area: zero;
  border-end-start-radius: var(--border-radius);
}

/* COLORS */

.white {
  background-color: var(--white);
}

.dark {
  background-color: var(--dark);
}

.green {
  background-color: var(--green);
}

.black {
  background-color: var(--black);
}

.grey {
  background-color: var(--grey);
}

.orange {
  background-color: var(--orange);
}

.shadow {
  box-shadow: 3px 3px 5px 0px var(--dark-grey);
}
