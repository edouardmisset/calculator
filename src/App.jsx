import { useEffect, useState } from 'react'
import './App.css'
import { calculate, concatenatePreviousValueAndNext } from './helpers'
import { StyledButton, StyledCalculator, StyledScreen } from './styles'
import {
  INDEX_TO_WORD,
  initialFirstNumber,
  initialSecondNumber,
  topRow,
  initialScreenValue,
} from './constants'

function App() {
  const [firstNumber, setFirstNumber] = useState(initialFirstNumber)
  const [secondNumber, setSecondNumber] = useState(initialSecondNumber)
  const [isFirstNumberSelected, setIsFirstNumberSelected] = useState(true)
  const [operator, setOperator] = useState()
  const [screenValue, setScreenValue] = useState(initialScreenValue)

  const changeSign = () => {
    if (isFirstNumberSelected) {
      setFirstNumber(previousValue => -previousValue)
    } else {
      setSecondNumber(previousValue => -previousValue)
    }
  }

  const handleNumberClick = event => {
    const value = event?.target?.value
    if (value === undefined) return
    const setNumber = concatenatePreviousValueAndNext(value)
    if (isFirstNumberSelected === true) {
      setFirstNumber(setNumber)
      setScreenValue(firstNumber)
    } else {
      setSecondNumber(setNumber)
      setScreenValue(secondNumber)
    }
  }

  const handleOperatorClick = event => {
    const op = event?.target?.value
    if (op === undefined) {
      console.log('No operator')
      return
    }
    setOperator(op)
    setIsFirstNumberSelected(state => !state)
  }

  const handleCalculate = () => {
    const result = calculate(
      Number(firstNumber),
      operator,
      Number(secondNumber),
    )
    setScreenValue(result)
    setSecondNumber('')
    setFirstNumber(result)
    setIsFirstNumberSelected(state => !state)
  }

  const handleClear = () => {
    setScreenValue('')
    setFirstNumber('')
    setSecondNumber('')
    setOperator()
    setIsFirstNumberSelected(true)
  }

  const functionButtons = [
    { position: 'inverse', value: '+/-', text: '+/-', handler: changeSign },
    { position: 'clear', value: 'C', text: 'C', handler: handleClear },
    { position: 'divide', value: '/', text: 'Ã·', handler: handleOperatorClick },
    {
      position: 'multiply',
      value: '*',
      text: 'x',
      handler: handleOperatorClick,
    },
    { position: 'add', value: '+', text: '+', handler: handleOperatorClick },
    {
      position: 'subtract',
      value: '-',
      text: '-',
      handler: handleOperatorClick,
    },
    { position: 'equal', value: '=', text: '=', handler: handleCalculate },
    { position: 'dot', value: '.', text: '.', handler: handleNumberClick },
  ]

  const numberButtons = INDEX_TO_WORD.map((text, index) => ({
    position: text,
    value: index,
    text: index,
    handler: handleNumberClick,
  }))

  const buttons = [...numberButtons, ...functionButtons]

  useEffect(() => {
    if (isFirstNumberSelected === true) {
      setScreenValue(firstNumber)
    } else {
      setScreenValue(secondNumber)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [firstNumber, secondNumber])

  return (
    <>
      <header>
        <h1>Awesome Calculator ðŸ”¥</h1>
      </header>
      <StyledCalculator>
        <StyledScreen value={screenValue} readOnly />
        {buttons.map(({ value, position, text, handler }) => (
          <StyledButton
            isZero={value === 0}
            isEqual={value === '='}
            onClick={handler}
            isTopRow={topRow.has(value)}
            key={value}
            position={position}
            value={value}
          >
            {text}
          </StyledButton>
        ))}
      </StyledCalculator>
    </>
  )
}

export default App
