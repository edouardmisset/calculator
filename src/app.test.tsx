import { render, screen } from '@testing-library/react'
import { describe, expect, it } from 'vitest'

import '@testing-library/jest-dom'
import App from './app'

describe('App', () => {
  it('Loads and displays the calculator', () => {
    // ARRANGE
    render(<App />)

    // ACT

    // ASSERT
    expect(screen.getByTestId('calculator')).toMatchSnapshot()
  })
})
