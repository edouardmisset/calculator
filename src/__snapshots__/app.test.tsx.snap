// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > Loads and displays the calculator 1`] = `
.c2 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--grey);
  border-end-start-radius: var(--border-radius);
  grid-area: zero;
}

.c2:hover {
  border-color: var(--main-color);
}

.c2:focus,
.c2 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c3 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--grey);
  grid-area: one;
}

.c3:hover {
  border-color: var(--main-color);
}

.c3:focus,
.c3 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c4 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--grey);
  grid-area: two;
}

.c4:hover {
  border-color: var(--main-color);
}

.c4:focus,
.c4 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c5 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--grey);
  grid-area: three;
}

.c5:hover {
  border-color: var(--main-color);
}

.c5:focus,
.c5 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c6 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--grey);
  grid-area: four;
}

.c6:hover {
  border-color: var(--main-color);
}

.c6:focus,
.c6 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c7 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--grey);
  grid-area: five;
}

.c7:hover {
  border-color: var(--main-color);
}

.c7:focus,
.c7 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c8 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--grey);
  grid-area: six;
}

.c8:hover {
  border-color: var(--main-color);
}

.c8:focus,
.c8 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c9 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--grey);
  grid-area: seven;
}

.c9:hover {
  border-color: var(--main-color);
}

.c9:focus,
.c9 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c10 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--grey);
  grid-area: eight;
}

.c10:hover {
  border-color: var(--main-color);
}

.c10:focus,
.c10 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c11 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--grey);
  grid-area: nine;
}

.c11:hover {
  border-color: var(--main-color);
}

.c11:focus,
.c11 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c12 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--black);
  grid-area: inverse;
}

.c12:hover {
  border-color: var(--main-color);
}

.c12:focus,
.c12 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c13 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--black);
  grid-area: clear;
}

.c13:hover {
  border-color: var(--main-color);
}

.c13:focus,
.c13 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c14 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--black);
  grid-area: divide;
}

.c14:hover {
  border-color: var(--main-color);
}

.c14:focus,
.c14 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c15 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--black);
  grid-area: multiply;
}

.c15:hover {
  border-color: var(--main-color);
}

.c15:focus,
.c15 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c16 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--grey);
  grid-area: add;
}

.c16:hover {
  border-color: var(--main-color);
}

.c16:focus,
.c16 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c17 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--grey);
  grid-area: subtract;
}

.c17:hover {
  border-color: var(--main-color);
}

.c17:focus,
.c17 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c18 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--main-color);
  border-end-end-radius: var(--border-radius);
  background-color: var(--main-color);
  grid-area: equal;
}

.c18:hover {
  border-color: var(--main-color);
}

.c18:focus,
.c18 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c19 {
  appearance: none;
  border: 1px solid transparent;
  border-radius: 0;
  color: var(--white);
  padding-inline: 0.5rem;
  padding-block: 1rem;
  font-size: 2em;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.3s;
  background-color: var(--grey);
  grid-area: dot;
}

.c19:hover {
  border-color: var(--main-color);
}

.c19:focus,
.c19 :focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.c1 {
  border: none;
  border-start-start-radius: var(--border-radius);
  border-start-end-radius: var(--border-radius);
  grid-area: screen;
  text-align: right;
  color: var(--black);
  font-size: 2rem;
  padding: 2rem;
  min-height: 4rem;
  font-family: monospace;
  background-color: var(--green);
}

.c0 {
  margin: auto;
  display: grid;
  grid-template-areas: 'screen screen screen screen' 'screen screen screen screen' 'clear inverse divide multiply' 'seven eight nine add' 'four five six subtract' 'one two three equal' 'zero zero dot equal';
  max-width: 500px;
  border-radius: var(--border-radius);
  box-shadow: 3px 3px 5px 0px var(--dark-grey);
  background-color: var(--dark);
}

<section
  class="c0"
  data-testid="calculator"
>
  <input
    class="c1"
    readonly=""
    value=""
  />
  <button
    class="c2"
    value="0"
  >
    0
  </button>
  <button
    class="c3"
    value="1"
  >
    1
  </button>
  <button
    class="c4"
    value="2"
  >
    2
  </button>
  <button
    class="c5"
    value="3"
  >
    3
  </button>
  <button
    class="c6"
    value="4"
  >
    4
  </button>
  <button
    class="c7"
    value="5"
  >
    5
  </button>
  <button
    class="c8"
    value="6"
  >
    6
  </button>
  <button
    class="c9"
    value="7"
  >
    7
  </button>
  <button
    class="c10"
    value="8"
  >
    8
  </button>
  <button
    class="c11"
    value="9"
  >
    9
  </button>
  <button
    class="c12"
    value="+/-"
  >
    +/-
  </button>
  <button
    class="c13"
    value="C"
  >
    C
  </button>
  <button
    class="c14"
    value="/"
  >
    รท
  </button>
  <button
    class="c15"
    value="*"
  >
    x
  </button>
  <button
    class="c16"
    value="+"
  >
    +
  </button>
  <button
    class="c17"
    value="-"
  >
    -
  </button>
  <button
    class="c18"
    value="="
  >
    =
  </button>
  <button
    class="c19"
    value="."
  >
    .
  </button>
</section>
`;
